clio-server/src/main/scala/org/broadinstitute/clio/server/service/IndexService.scala

```diff
@@ -67,4 +78,133 @@ abstract class IndexService[CI <: ClioIndex](
+      ElasticsearchIndex.EntityIdElasticsearchName,
+    )
+
+  def validateUpsert(
+    indexKey: clioIndex.KeyType,
+    metadata: clioIndex.MetadataType,
+    force: Boolean
+  )(
+    implicit materializer: Materializer
+  ): Either[Exception, clioIndex.MetadataType] = {
+
+    // drop housekeeping and key fields (evaluate only metadata)
+    val dropKeys = keysToDrop ++ indexKey.getClass.getDeclaredFields
+      .map(field => ElasticsearchUtil.toElasticsearchName(field.getName))
+
+    val storedDocsForKey = Await
+      .result(
+        searchService
+          .simpleStringQueryMetadata(indexKey.asJson.pretty(Printer.noSpaces))(
+            elasticsearchIndex
+          )
+          .runWith(Sink.seq),
+        1.second
+      )
+      .map(
+        json =>
+          json.mapObject(_.filterKeys({ key =>
+            !dropKeys.contains(key)
+          }))
+      )
+
+    if (storedDocsForKey.size > 1) {
+      Left(
+        UpsertValidationException(
+          s"Key based query returned more than one entry. $indexKey"
+        )
+      )
+    }
+
+    if (storedDocsForKey.isEmpty) {
+      // if there are no documents stored set document status to either what was passed in or Normal (default)
+      Right(
+        metadata.documentStatus
+          .fold(metadata.withDocumentStatus(Some(Normal)))(_ => metadata)
+      )
+    } else {
+      // validate that we will not overwrite any data?
+      val storedDoc = storedDocsForKey.head.as[clioIndex.MetadataType].toOption
+
+      val updatedMetadata = checkStoredDocumentStatus(metadata, storedDoc)
+      val differences = storedDoc.fold(Iterable.empty[(String, Any, Any)]) {
+        diffMetadata(_, updatedMetadata)
+      }
+      if (!force && differences.nonEmpty) {
+        val diffs = differences.map {
+          case (field, newVal, oldVal) =>
+            s"Field: $field, Old value: $oldVal, New value: $newVal"
+        }.mkString("\n")
+
+        Left(
+          UpsertValidationException(
+            s"""Adding this document will overwrite the following existing metadata:
+$diffs
+
+Use 'force=true' to overwrite the existing data.""".stripMargin
+          )
+        )
+      } else { Right(updatedMetadata) }
+    }
+  }
+
+  def checkStoredDocumentStatus(
+    metadata: clioIndex.MetadataType,
+    storedMetadata: Option[clioIndex.MetadataType]
+  ): clioIndex.MetadataType = {
+    val updatedMetadata = storedMetadata.fold(
+      // if there is no metadata currently stored for this key then set document status to Normal
+      metadata.withDocumentStatus(Some(DocumentStatus.Normal))
+    )(
+      meta =>
+        meta.documentStatus.fold(
+          // if there is metadata stored for this key but no document status (shouldn't happen) set it to Normal
+          metadata.withDocumentStatus(Some(Normal))
+        )(
+          // otherwise use the metadata passed in
+          _ => metadata
+      )
+    )
+
+    updatedMetadata
+  }
+
+  private def diffMetadata(
```

Comment:
It looks like this is a static method. If so, maybe it should be moved to a class more specific to its definition, such as `ClioIndex`. That way if we ever move this check again, we won't have to move this diff operation again.